{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fcdaec30-0d07-4a80-aa32-21c4da1143be",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages(library(GenomicFeatures))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0cfe5cc5-677b-443b-942e-dd143eb9c689",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages(library(systemPipeR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9663b7be-d0c5-400f-bc9a-ef11d47b5791",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8d653f24-9c6f-4034-baeb-48812007607d",
   "metadata": {},
   "outputs": [],
   "source": [
    "exonsbygene <- exonsBy(txdb, by=\"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "33e98d76-fc78-4696-b766-d9f8167a8804",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'CompressedGRangesList'"
      ],
      "text/latex": [
       "'CompressedGRangesList'"
      ],
      "text/markdown": [
       "'CompressedGRangesList'"
      ],
      "text/plain": [
       "[1] \"CompressedGRangesList\"\n",
       "attr(,\"package\")\n",
       "[1] \"GenomicRanges\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class(exonsbygene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "843db29c-bdab-4d90-b9ab-b2de46316bb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for seqinfo {GenomeInfoDb}\"><tr><td>seqinfo {GenomeInfoDb}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Accessing/modifying sequence information</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>A set of generic functions for getting/setting/modifying the sequence\n",
       "information stored in an object.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "seqinfo(x)\n",
       "seqinfo(x,\n",
       "        new2old=NULL,\n",
       "        pruning.mode=c(\"error\", \"coarse\", \"fine\", \"tidy\")) &lt;- value\n",
       "\n",
       "seqnames(x)\n",
       "seqnames(x) &lt;- value\n",
       "\n",
       "seqlevels(x)\n",
       "seqlevels(x,\n",
       "          pruning.mode=c(\"error\", \"coarse\", \"fine\", \"tidy\")) &lt;- value\n",
       "sortSeqlevels(x, X.is.sexchrom=NA)\n",
       "seqlevelsInUse(x)\n",
       "seqlevels0(x)\n",
       "\n",
       "seqlengths(x)\n",
       "seqlengths(x) &lt;- value\n",
       "\n",
       "isCircular(x)\n",
       "isCircular(x) &lt;- value\n",
       "\n",
       "genome(x)\n",
       "genome(x) &lt;- value\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>x</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Any object containing sequence information i.e. with a <code>seqinfo()</code>\n",
       "component.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>new2old</code></td>\n",
       "<td>\n",
       "\n",
       "<p>The <code>new2old</code> argument allows the user to rename, drop, add and/or\n",
       "reorder the &quot;sequence levels&quot; in <code>x</code>.\n",
       "</p>\n",
       "<p><code>new2old</code> can be <code>NULL</code> or an integer vector with one element\n",
       "per entry in Seqinfo object <code>value</code> (i.e. <code>new2old</code> and\n",
       "<code>value</code> must have the same length) describing how the &quot;new&quot; sequence\n",
       "levels should be mapped to the &quot;old&quot; sequence levels, that is, how the\n",
       "entries in <code>value</code> should be mapped to the entries in\n",
       "<code>seqinfo(x)</code>.\n",
       "The values in <code>new2old</code> must be &gt;= 1 and &lt;= <code>length(seqinfo(x))</code>.\n",
       "<code>NA</code>s are allowed and indicate sequence levels that are being added.\n",
       "Old sequence levels that are not represented in <code>new2old</code> will be\n",
       "dropped, but this will fail if those levels are in use (e.g. if <code>x</code>\n",
       "is a GRanges object with ranges defined on those\n",
       "sequence levels) unless a pruning mode is specified via the\n",
       "<code>pruning.mode</code> argument (see below).\n",
       "</p>\n",
       "<p>If <code>new2old=NULL</code>, then sequence levels can only be added to the\n",
       "existing ones, that is, <code>value</code> must have at least as many entries\n",
       "as <code>seqinfo(x)</code> (i.e. <code>length(values) &gt;= length(seqinfo(x))</code>)\n",
       "and also <code>seqlevels(values)[seq_len(length(seqlevels(x)))]</code> must be\n",
       "identical to <code>seqlevels(x)</code>.\n",
       "</p>\n",
       "<p>Note that most of the times it's easier to proceed in 2 steps:\n",
       "</p>\n",
       "\n",
       "<ol>\n",
       "<li><p> First align the seqlevels on the left (<code>seqlevels(x)</code>)\n",
       "with the seqlevels on the right.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> Then call <code>seqinfo(x) &lt;- value</code>. Because\n",
       "<code>seqlevels(x)</code> and <code>seqlevels(value)</code> now are\n",
       "identical, there's no need to specify <code>new2old</code>.\n",
       "</p>\n",
       "</li></ol>\n",
       "\n",
       "<p>This 2-step approach will typically look like this:\n",
       "</p>\n",
       "<pre>    seqlevels(x) &lt;- seqlevels(value)  # align seqlevels\n",
       "    seqinfo(x) &lt;- seqinfo(value)  # guaranteed to work\n",
       "    </pre>\n",
       "<p>Or, if <code>x</code> has seqlevels not in <code>value</code>, it will look\n",
       "like this:\n",
       "</p>\n",
       "<pre>    seqlevels(x, pruning.mode=\"coarse\") &lt;- seqlevels(value)\n",
       "    seqinfo(x) &lt;- seqinfo(value)  # guaranteed to work\n",
       "    </pre>\n",
       "<p>The <code>pruning.mode</code> argument will control what happens to <code>x</code>\n",
       "when some of its seqlevels get droppped. See below for more information.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pruning.mode</code></td>\n",
       "<td>\n",
       "\n",
       "<p>When some of the seqlevels to drop from <code>x</code> are in use (i.e. have\n",
       "ranges on them), the ranges on these sequences need to be removed before\n",
       "the seqlevels can be dropped. We call this <em>pruning</em>. The\n",
       "<code>pruning.mode</code> argument controls how to <em>prune</em> <code>x</code>.\n",
       "Four pruning modes are currently defined: <code>\"error\"</code>, <code>\"coarse\"</code>,\n",
       "<code>\"fine\"</code>, and <code>\"tidy\"</code>. <code>\"error\"</code> is the default. In this\n",
       "mode, no pruning is done and an error is raised. The other pruning modes\n",
       "do the following:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>\"coarse\"</code>: Remove the elements in <code>x</code> where the\n",
       "seqlevels to drop are in use. Typically reduces the length of\n",
       "<code>x</code>. Note that if <code>x</code> is a list-like object (e.g.\n",
       "GRangesList,\n",
       "GAlignmentPairs, or\n",
       "GAlignmentsList), then any list\n",
       "element in <code>x</code> where at least one of the sequence levels\n",
       "to drop is in use is <em>fully</em> removed. In other words,\n",
       "when <code>pruning.mode=\"coarse\"</code>, the <code>seqlevels</code> setter\n",
       "will keep or remove <em>full list elements</em> and not try to\n",
       "change their content. This guarantees that the exact ranges\n",
       "(and their order) inside the individual list elements are\n",
       "preserved. This can be a desirable property when the list\n",
       "elements represent compound features like exons grouped by\n",
       "transcript (stored in a GRangesList\n",
       "object as returned by\n",
       "<code>exonsBy( , by=\"tx\")</code>),\n",
       "or paired-end or fusion reads, etc...\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>\"fine\"</code>: Supported on list-like objects only. Removes\n",
       "the ranges that are on the sequences to drop. This removal is\n",
       "done within each list element of the original object <code>x</code>\n",
       "and doesn't affect its length or the order of its list elements.\n",
       "In other words, the pruned object is guaranteed to be\n",
       "<em>parallel</em> to the original object.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>\"tidy\"</code>: Like the <code>\"fine\"</code> pruning above but also\n",
       "removes the list elements that become empty as the result of\n",
       "the pruning. Note that this pruning mode is particularly well\n",
       "suited on a GRangesList object that\n",
       "contains transcripts grouped by gene, as returned by\n",
       "<code>transcriptsBy( , by=\"gene\")</code>.\n",
       "Finally note that, as a convenience, this pruning mode is\n",
       "supported on non list-like objects (e.g.\n",
       "GRanges or\n",
       "GAlignments objects) and, in this\n",
       "case, is equivalent to the <code>\"coarse\"</code> mode.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "<p>See the &quot;B. DROP SEQLEVELS FROM A LIST-LIKE OBJECT&quot; section in the examples\n",
       "below for an extensive illustration of these pruning modes.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>value</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Typically a Seqinfo object for the <code>seqinfo</code> setter.\n",
       "</p>\n",
       "<p>Either a named or unnamed character vector for the <code>seqlevels</code>\n",
       "setter.\n",
       "</p>\n",
       "<p>A vector containing the sequence information to store for the other\n",
       "setters.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>X.is.sexchrom</code></td>\n",
       "<td>\n",
       "\n",
       "<p>A logical indicating whether X refers to the sexual chromosome\n",
       "or to chromosome with Roman Numeral X. If <code>NA</code>, <code>sortSeqlevels</code>\n",
       "does its best to &quot;guess&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>It all revolves around Seqinfo objects</h3>\n",
       "\n",
       "<p>The Seqinfo class plays a central role for the functions described\n",
       "in this man page because:\n",
       "</p>\n",
       "\n",
       "<ol>\n",
       "<li><p> All these functions (except <code>seqinfo</code>, <code>seqlevelsInUse</code>,\n",
       "and <code>seqlevels0</code>) work on a Seqinfo object.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> For classes that implement it, the <code>seqinfo</code> getter should\n",
       "return a Seqinfo object.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> Default <code>seqlevels</code>, <code>seqlengths</code>, <code>isCircular</code>,\n",
       "and <code>genome</code> getters and setters are provided.\n",
       "By default, <code>seqlevels(x)</code> does <code>seqlevels(seqinfo(x))</code>,\n",
       "<code>seqlengths(x)</code> does <code>seqlengths(seqinfo(x))</code>,\n",
       "<code>isCircular(x)</code> does <code>isCircular(seqinfo(x))</code>,\n",
       "and <code>genome(x)</code> does <code>genome(seqinfo(x))</code>.\n",
       "So any class with a <code>seqinfo</code> getter will have all the above\n",
       "getters work out-of-the-box. If, in addition, the class defines\n",
       "a <code>seqinfo</code> setter, then all the corresponding setters will\n",
       "also work out-of-the-box.\n",
       "</p>\n",
       "<p>Examples of containers that have a <code>seqinfo</code> getter\n",
       "and setter:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> the GRanges and\n",
       "GRangesList classes\n",
       "in the <span class=\"pkg\">GenomicRanges</span> package;\n",
       "</p>\n",
       "</li>\n",
       "<li><p> the SummarizedExperiment\n",
       "class in the <span class=\"pkg\">SummarizedExperiment</span> package;\n",
       "</p>\n",
       "</li>\n",
       "<li><p> the GAlignments,\n",
       "GAlignmentPairs,\n",
       "and GAlignmentsList classes\n",
       "in the <span class=\"pkg\">GenomicAlignments</span> package;\n",
       "</p>\n",
       "</li>\n",
       "<li><p> the TxDb class in the\n",
       "<span class=\"pkg\">GenomicFeatures</span> package;\n",
       "</p>\n",
       "</li>\n",
       "<li><p> the BSgenome class in the\n",
       "<span class=\"pkg\">BSgenome</span> package;\n",
       "</p>\n",
       "</li>\n",
       "<li><p> and more...\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li></ol>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Note</h3>\n",
       "\n",
       "<p>The full list of methods defined for a given generic function can be seen\n",
       "with e.g. <code>showMethods(\"seqinfo\")</code> or <code>showMethods(\"seqnames\")</code>\n",
       "(for the getters), and <code>showMethods(\"seqinfo&lt;-\")</code> or\n",
       "<code>showMethods(\"seqnames&lt;-\")</code> (for the setters a.k.a.\n",
       "<em>replacement methods</em>). Please be aware that this shows only methods\n",
       "defined in packages that are currently attached.\n",
       "</p>\n",
       "<p>The <span class=\"pkg\">GenomicRanges</span> package defines <code>seqinfo</code> and <code>seqinfo&lt;-</code>\n",
       "methods for these low-level data types: List and\n",
       "IntegerRangesList. Those objects do not have the means\n",
       "to formally store sequence information. Thus, the wrappers simply store\n",
       "the Seqinfo object within <code>metadata(x)</code>. Initially, the\n",
       "metadata is empty, so there is some effort to generate a reasonable\n",
       "default Seqinfo. The names of any List are\n",
       "taken as the <code>seqnames</code>, and the <code>universe</code> of\n",
       "IntegerRangesList is taken as the <code>genome</code>.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>H. Pagès</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "\n",
       "<ul>\n",
       "<li><p> The seqlevelsStyle generic getter and setter for\n",
       "conveniently renaming the seqlevels of an object according\n",
       "to a given naming convention (e.g. NCBI or UCSC).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>Seqinfo objects.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>GRanges and GRangesList\n",
       "objects in the <span class=\"pkg\">GenomicRanges</span> package.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>SummarizedExperiment objects in the\n",
       "<span class=\"pkg\">SummarizedExperiment</span> package.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>GAlignments,\n",
       "GAlignmentPairs,\n",
       "and GAlignmentsList objects in the\n",
       "<span class=\"pkg\">GenomicAlignments</span> package.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>TxDb objects in the\n",
       "<span class=\"pkg\">GenomicFeatures</span> package.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>BSgenome objects in the <span class=\"pkg\">BSgenome</span> package.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>seqlevels-wrappers for convenience wrappers to the\n",
       "<code>seqlevels</code> getter and setter.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>rankSeqlevels</code>, on which <code>sortSeqlevels</code> is\n",
       "based.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "## ---------------------------------------------------------------------\n",
       "## A. BASIC USAGE OF THE seqlevels() GETTER AND SETTER\n",
       "## ---------------------------------------------------------------------\n",
       "## Operations between 2 or more objects containing genomic ranges (e.g.\n",
       "## finding overlaps, comparing, or matching) only make sense if the\n",
       "## operands have the same seqlevels. So before performing such\n",
       "## operations, it is often necessary to adjust the seqlevels in\n",
       "## the operands so that they all have the same seqlevels. This is\n",
       "## typically done with the seqlevels() setter. The setter can be used\n",
       "## to rename, drop, add and/or reorder seqlevels of an object. The\n",
       "## examples below show how to mofify the seqlevels of a GRanges object\n",
       "## but the same would apply to any object containing sequence\n",
       "## information (i.e. with a seqinfo() component).\n",
       "library(GenomicRanges)\n",
       "\n",
       "gr &lt;- GRanges(rep(c(\"chr2\", \"chr3\", \"chrM\"), 2), IRanges(1:6, 10))\n",
       "\n",
       "## Add new seqlevels:\n",
       "seqlevels(gr) &lt;- c(\"chr1\", seqlevels(gr), \"chr4\")\n",
       "seqlevels(gr)\n",
       "seqlevelsInUse(gr)\n",
       "\n",
       "## Reorder existing seqlevels:\n",
       "seqlevels(gr) &lt;- rev(seqlevels(gr))\n",
       "seqlevels(gr)\n",
       "\n",
       "## Drop all unused seqlevels:\n",
       "seqlevels(gr) &lt;- seqlevelsInUse(gr)\n",
       "\n",
       "## Drop some seqlevels in use:\n",
       "seqlevels(gr, pruning.mode=\"coarse\") &lt;- setdiff(seqlevels(gr), \"chr3\")\n",
       "gr\n",
       "\n",
       "## Rename, add, and reorder the seqlevels all at once:\n",
       "seqlevels(gr) &lt;- c(\"chr1\", chr2=\"chr2\", chrM=\"Mitochondrion\")\n",
       "seqlevels(gr)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## B. DROP SEQLEVELS FROM A LIST-LIKE OBJECT\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "grl0 &lt;- GRangesList(A=GRanges(\"chr2\", IRanges(3:2, 5)),\n",
       "                    B=GRanges(c(\"chr2\", \"chrMT\"), IRanges(7:6, 15)),\n",
       "                    C=GRanges(c(\"chrY\", \"chrMT\"), IRanges(17:16, 25)),\n",
       "                    D=GRanges())\n",
       "grl0\n",
       "\n",
       "grl1 &lt;- grl0\n",
       "seqlevels(grl1, pruning.mode=\"coarse\") &lt;- c(\"chr2\", \"chr5\")\n",
       "grl1  # grl0[[2]] was fully removed! (even if it had a range on chr2)\n",
       "\n",
       "## If what is desired is to remove the 2nd range in grl0[[2]] only (i.e.\n",
       "## the chrMT:6-15 range), or, more generally speaking, to remove the\n",
       "## ranges within each list element that are located on the seqlevels to\n",
       "## drop, then use pruning.mode=\"fine\" or pruning.mode=\"tidy\":\n",
       "grl2 &lt;- grl0\n",
       "seqlevels(grl2, pruning.mode=\"fine\") &lt;- c(\"chr2\", \"chr5\")\n",
       "grl2  # grl0[[2]] not removed, but chrMT:6-15 range removed from it\n",
       "\n",
       "## Like pruning.mode=\"fine\" but also removes grl0[[3]].\n",
       "grl3 &lt;- grl0\n",
       "seqlevels(grl3, pruning.mode=\"tidy\") &lt;- c(\"chr2\", \"chr5\")\n",
       "grl3\n",
       "\n",
       "library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)\n",
       "txdb &lt;- TxDb.Dmelanogaster.UCSC.dm3.ensGene\n",
       "## Pruning mode \"coarse\" is particularly well suited on a GRangesList\n",
       "## object that contains exons grouped by transcript:\n",
       "ex_by_tx &lt;- exonsBy(txdb, by=\"tx\")\n",
       "seqlevels(ex_by_tx)\n",
       "seqlevels(ex_by_tx, pruning.mode=\"coarse\") &lt;- \"chr2L\"\n",
       "seqlevels(ex_by_tx)\n",
       "## Pruning mode \"tidy\" is particularly well suited on a GRangesList\n",
       "## object that contains transcripts grouped by gene:\n",
       "tx_by_gene &lt;- transcriptsBy(txdb, by=\"gene\")\n",
       "seqlevels(tx_by_gene)\n",
       "seqlevels(tx_by_gene, pruning.mode=\"tidy\") &lt;- \"chr2L\"\n",
       "seqlevels(tx_by_gene)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## C. RENAME THE SEQLEVELS OF A TxDb OBJECT\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)\n",
       "txdb &lt;- TxDb.Dmelanogaster.UCSC.dm3.ensGene\n",
       "seqlevels(txdb)\n",
       "\n",
       "seqlevels(txdb) &lt;- sub(\"chr\", \"\", seqlevels(txdb))\n",
       "seqlevels(txdb)\n",
       "\n",
       "seqlevels(txdb) &lt;- paste0(\"CH\", seqlevels(txdb))\n",
       "seqlevels(txdb)\n",
       "\n",
       "seqlevels(txdb)[seqlevels(txdb) == \"CHM\"] &lt;- \"M\"\n",
       "seqlevels(txdb)\n",
       "\n",
       "## Restore original seqlevels:\n",
       "seqlevels(txdb) &lt;- seqlevels0(txdb)\n",
       "seqlevels(txdb)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## D. SORT SEQLEVELS IN \"NATURAL\" ORDER\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "sortSeqlevels(c(\"11\", \"Y\", \"1\", \"10\", \"9\", \"M\", \"2\"))\n",
       "\n",
       "seqlevels &lt;- c(\"chrXI\", \"chrY\", \"chrI\", \"chrX\", \"chrIX\", \"chrM\", \"chrII\")\n",
       "sortSeqlevels(seqlevels)\n",
       "sortSeqlevels(seqlevels, X.is.sexchrom=TRUE)\n",
       "sortSeqlevels(seqlevels, X.is.sexchrom=FALSE)\n",
       "\n",
       "seqlevels &lt;- c(\"chr2RHet\", \"chr4\", \"chrUextra\", \"chrYHet\",\n",
       "               \"chrM\", \"chrXHet\", \"chr2LHet\", \"chrU\",\n",
       "               \"chr3L\", \"chr3R\", \"chr2R\", \"chrX\")\n",
       "sortSeqlevels(seqlevels)\n",
       "\n",
       "gr &lt;- GRanges()\n",
       "seqlevels(gr) &lt;- seqlevels\n",
       "sortSeqlevels(gr)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## E. SUBSET OBJECTS BY SEQLEVELS\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "tx &lt;- transcripts(txdb)\n",
       "seqlevels(tx)\n",
       "\n",
       "## Drop 'M', keep all others.\n",
       "seqlevels(tx, pruning.mode=\"coarse\") &lt;- seqlevels(tx)[seqlevels(tx) != \"M\"]\n",
       "seqlevels(tx)\n",
       "\n",
       "## Drop all except 'ch3L' and 'ch3R'.\n",
       "seqlevels(tx, pruning.mode=\"coarse\") &lt;- c(\"ch3L\", \"ch3R\")\n",
       "seqlevels(tx)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## F. FINDING METHODS\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "showMethods(\"seqinfo\")\n",
       "showMethods(\"seqinfo&lt;-\")\n",
       "\n",
       "showMethods(\"seqnames\")\n",
       "showMethods(\"seqnames&lt;-\")\n",
       "\n",
       "showMethods(\"seqlevels\")\n",
       "showMethods(\"seqlevels&lt;-\")\n",
       "\n",
       "if (interactive()) {\n",
       "  library(GenomicRanges)\n",
       "  ?`GRanges-class`\n",
       "}\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>GenomeInfoDb</em> version 1.30.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{seqinfo}{Accessing/modifying sequence information}{seqinfo}\n",
       "\\aliasA{genome}{seqinfo}{genome}\n",
       "\\aliasA{genome,ANY-method}{seqinfo}{genome,ANY.Rdash.method}\n",
       "\\aliasA{genome<\\Rdash{}}{seqinfo}{genome<.Rdash.}\n",
       "\\aliasA{genome<\\Rdash{},ANY-method}{seqinfo}{genome<.Rdash.,ANY.Rdash.method}\n",
       "\\aliasA{isCircular}{seqinfo}{isCircular}\n",
       "\\aliasA{isCircular,ANY-method}{seqinfo}{isCircular,ANY.Rdash.method}\n",
       "\\aliasA{isCircular<\\Rdash{}}{seqinfo}{isCircular<.Rdash.}\n",
       "\\aliasA{isCircular<\\Rdash{},ANY-method}{seqinfo}{isCircular<.Rdash.,ANY.Rdash.method}\n",
       "\\aliasA{seqinfo<\\Rdash{}}{seqinfo}{seqinfo<.Rdash.}\n",
       "\\aliasA{seqlengths}{seqinfo}{seqlengths}\n",
       "\\aliasA{seqlengths,ANY-method}{seqinfo}{seqlengths,ANY.Rdash.method}\n",
       "\\aliasA{seqlengths<\\Rdash{}}{seqinfo}{seqlengths<.Rdash.}\n",
       "\\aliasA{seqlengths<\\Rdash{},ANY-method}{seqinfo}{seqlengths<.Rdash.,ANY.Rdash.method}\n",
       "\\aliasA{seqlevels}{seqinfo}{seqlevels}\n",
       "\\aliasA{seqlevels,ANY-method}{seqinfo}{seqlevels,ANY.Rdash.method}\n",
       "\\aliasA{seqlevels<\\Rdash{}}{seqinfo}{seqlevels<.Rdash.}\n",
       "\\aliasA{seqlevels<\\Rdash{},ANY-method}{seqinfo}{seqlevels<.Rdash.,ANY.Rdash.method}\n",
       "\\aliasA{seqlevels0}{seqinfo}{seqlevels0}\n",
       "\\aliasA{seqlevelsInUse}{seqinfo}{seqlevelsInUse}\n",
       "\\aliasA{seqlevelsInUse,CompressedList-method}{seqinfo}{seqlevelsInUse,CompressedList.Rdash.method}\n",
       "\\aliasA{seqlevelsInUse,Vector-method}{seqinfo}{seqlevelsInUse,Vector.Rdash.method}\n",
       "\\aliasA{seqnames}{seqinfo}{seqnames}\n",
       "\\aliasA{seqnames<\\Rdash{}}{seqinfo}{seqnames<.Rdash.}\n",
       "\\aliasA{sortSeqlevels}{seqinfo}{sortSeqlevels}\n",
       "\\aliasA{sortSeqlevels,ANY-method}{seqinfo}{sortSeqlevels,ANY.Rdash.method}\n",
       "\\aliasA{sortSeqlevels,character-method}{seqinfo}{sortSeqlevels,character.Rdash.method}\n",
       "\\keyword{methods}{seqinfo}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "A set of generic functions for getting/setting/modifying the sequence\n",
       "information stored in an object.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "seqinfo(x)\n",
       "seqinfo(x,\n",
       "        new2old=NULL,\n",
       "        pruning.mode=c(\"error\", \"coarse\", \"fine\", \"tidy\")) <- value\n",
       "\n",
       "seqnames(x)\n",
       "seqnames(x) <- value\n",
       "\n",
       "seqlevels(x)\n",
       "seqlevels(x,\n",
       "          pruning.mode=c(\"error\", \"coarse\", \"fine\", \"tidy\")) <- value\n",
       "sortSeqlevels(x, X.is.sexchrom=NA)\n",
       "seqlevelsInUse(x)\n",
       "seqlevels0(x)\n",
       "\n",
       "seqlengths(x)\n",
       "seqlengths(x) <- value\n",
       "\n",
       "isCircular(x)\n",
       "isCircular(x) <- value\n",
       "\n",
       "genome(x)\n",
       "genome(x) <- value\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{x}] \n",
       "Any object containing sequence information i.e. with a \\code{seqinfo()}\n",
       "component.\n",
       "\n",
       "\\item[\\code{new2old}] \n",
       "The \\code{new2old} argument allows the user to rename, drop, add and/or\n",
       "reorder the \"sequence levels\" in \\code{x}.\n",
       "\n",
       "\\code{new2old} can be \\code{NULL} or an integer vector with one element\n",
       "per entry in \\LinkA{Seqinfo}{Seqinfo} object \\code{value} (i.e. \\code{new2old} and\n",
       "\\code{value} must have the same length) describing how the \"new\" sequence\n",
       "levels should be mapped to the \"old\" sequence levels, that is, how the\n",
       "entries in \\code{value} should be mapped to the entries in\n",
       "\\code{seqinfo(x)}.\n",
       "The values in \\code{new2old} must be >= 1 and <= \\code{length(seqinfo(x))}.\n",
       "\\code{NA}s are allowed and indicate sequence levels that are being added.\n",
       "Old sequence levels that are not represented in \\code{new2old} will be\n",
       "dropped, but this will fail if those levels are in use (e.g. if \\code{x}\n",
       "is a \\LinkA{GRanges}{GRanges} object with ranges defined on those\n",
       "sequence levels) unless a pruning mode is specified via the\n",
       "\\code{pruning.mode} argument (see below).\n",
       "\n",
       "If \\code{new2old=NULL}, then sequence levels can only be added to the\n",
       "existing ones, that is, \\code{value} must have at least as many entries\n",
       "as \\code{seqinfo(x)} (i.e. \\code{length(values) >= length(seqinfo(x))})\n",
       "and also \\code{seqlevels(values)[seq\\_len(length(seqlevels(x)))]} must be\n",
       "identical to \\code{seqlevels(x)}.\n",
       "\n",
       "Note that most of the times it's easier to proceed in 2 steps:\n",
       "\\begin{enumerate}\n",
       "\n",
       "\\item{} First align the seqlevels on the left (\\code{seqlevels(x)})\n",
       "with the seqlevels on the right.\n",
       "\\item{} Then call \\code{seqinfo(x) <- value}. Because\n",
       "\\code{seqlevels(x)} and \\code{seqlevels(value)} now are\n",
       "identical, there's no need to specify \\code{new2old}.\n",
       "\n",
       "\\end{enumerate}\n",
       "\n",
       "This 2-step approach will typically look like this:\n",
       "\\begin{alltt}    seqlevels(x) <- seqlevels(value)  # align seqlevels\n",
       "    seqinfo(x) <- seqinfo(value)  # guaranteed to work\n",
       "    \\end{alltt}\n",
       "\n",
       "Or, if \\code{x} has seqlevels not in \\code{value}, it will look\n",
       "like this:\n",
       "\\begin{alltt}    seqlevels(x, pruning.mode=\"coarse\") <- seqlevels(value)\n",
       "    seqinfo(x) <- seqinfo(value)  # guaranteed to work\n",
       "    \\end{alltt}\n",
       "\n",
       "The \\code{pruning.mode} argument will control what happens to \\code{x}\n",
       "when some of its seqlevels get droppped. See below for more information.\n",
       "\n",
       "\\item[\\code{pruning.mode}] \n",
       "When some of the seqlevels to drop from \\code{x} are in use (i.e. have\n",
       "ranges on them), the ranges on these sequences need to be removed before\n",
       "the seqlevels can be dropped. We call this \\emph{pruning}. The\n",
       "\\code{pruning.mode} argument controls how to \\emph{prune} \\code{x}.\n",
       "Four pruning modes are currently defined: \\code{\"error\"}, \\code{\"coarse\"},\n",
       "\\code{\"fine\"}, and \\code{\"tidy\"}. \\code{\"error\"} is the default. In this\n",
       "mode, no pruning is done and an error is raised. The other pruning modes\n",
       "do the following:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} \\code{\"coarse\"}: Remove the elements in \\code{x} where the\n",
       "seqlevels to drop are in use. Typically reduces the length of\n",
       "\\code{x}. Note that if \\code{x} is a list-like object (e.g.\n",
       "\\LinkA{GRangesList}{GRangesList},\n",
       "\\LinkA{GAlignmentPairs}{GAlignmentPairs}, or\n",
       "\\LinkA{GAlignmentsList}{GAlignmentsList}), then any list\n",
       "element in \\code{x} where at least one of the sequence levels\n",
       "to drop is in use is \\emph{fully} removed. In other words,\n",
       "when \\code{pruning.mode=\"coarse\"}, the \\code{seqlevels} setter\n",
       "will keep or remove \\emph{full list elements} and not try to\n",
       "change their content. This guarantees that the exact ranges\n",
       "(and their order) inside the individual list elements are\n",
       "preserved. This can be a desirable property when the list\n",
       "elements represent compound features like exons grouped by\n",
       "transcript (stored in a \\LinkA{GRangesList}{GRangesList}\n",
       "object as returned by\n",
       "\\code{\\LinkA{exonsBy}{exonsBy}( , by=\"tx\")}),\n",
       "or paired-end or fusion reads, etc...\n",
       "\n",
       "\\item{} \\code{\"fine\"}: Supported on list-like objects only. Removes\n",
       "the ranges that are on the sequences to drop. This removal is\n",
       "done within each list element of the original object \\code{x}\n",
       "and doesn't affect its length or the order of its list elements.\n",
       "In other words, the pruned object is guaranteed to be\n",
       "\\emph{parallel} to the original object.\n",
       "\n",
       "\\item{} \\code{\"tidy\"}: Like the \\code{\"fine\"} pruning above but also\n",
       "removes the list elements that become empty as the result of\n",
       "the pruning. Note that this pruning mode is particularly well\n",
       "suited on a \\LinkA{GRangesList}{GRangesList} object that\n",
       "contains transcripts grouped by gene, as returned by\n",
       "\\code{\\LinkA{transcriptsBy}{transcriptsBy}( , by=\"gene\")}.\n",
       "Finally note that, as a convenience, this pruning mode is\n",
       "supported on non list-like objects (e.g.\n",
       "\\LinkA{GRanges}{GRanges} or\n",
       "\\LinkA{GAlignments}{GAlignments} objects) and, in this\n",
       "case, is equivalent to the \\code{\"coarse\"} mode.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "See the \"B. DROP SEQLEVELS FROM A LIST-LIKE OBJECT\" section in the examples\n",
       "below for an extensive illustration of these pruning modes.\n",
       "\n",
       "\\item[\\code{value}] \n",
       "Typically a \\LinkA{Seqinfo}{Seqinfo} object for the \\code{seqinfo} setter.\n",
       "\n",
       "Either a named or unnamed character vector for the \\code{seqlevels}\n",
       "setter.\n",
       "\n",
       "A vector containing the sequence information to store for the other\n",
       "setters.\n",
       "\n",
       "\\item[\\code{X.is.sexchrom}] \n",
       "A logical indicating whether X refers to the sexual chromosome\n",
       "or to chromosome with Roman Numeral X. If \\code{NA}, \\code{sortSeqlevels}\n",
       "does its best to \"guess\".\n",
       "\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Section}{It all revolves around Seqinfo objects}\n",
       "The \\LinkA{Seqinfo}{Seqinfo} class plays a central role for the functions described\n",
       "in this man page because:\n",
       "\n",
       "\\begin{enumerate}\n",
       "\n",
       "\\item{} All these functions (except \\code{seqinfo}, \\code{seqlevelsInUse},\n",
       "and \\code{seqlevels0}) work on a \\LinkA{Seqinfo}{Seqinfo} object.\n",
       "\n",
       "\\item{} For classes that implement it, the \\code{seqinfo} getter should\n",
       "return a \\LinkA{Seqinfo}{Seqinfo} object.\n",
       "\n",
       "\\item{} Default \\code{seqlevels}, \\code{seqlengths}, \\code{isCircular},\n",
       "and \\code{genome} getters and setters are provided.\n",
       "By default, \\code{seqlevels(x)} does \\code{seqlevels(seqinfo(x))},\n",
       "\\code{seqlengths(x)} does \\code{seqlengths(seqinfo(x))},\n",
       "\\code{isCircular(x)} does \\code{isCircular(seqinfo(x))},\n",
       "and \\code{genome(x)} does \\code{genome(seqinfo(x))}.\n",
       "So any class with a \\code{seqinfo} getter will have all the above\n",
       "getters work out-of-the-box. If, in addition, the class defines\n",
       "a \\code{seqinfo} setter, then all the corresponding setters will\n",
       "also work out-of-the-box.\n",
       "\n",
       "Examples of containers that have a \\code{seqinfo} getter\n",
       "and setter:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} the \\LinkA{GRanges}{GRanges} and\n",
       "\\LinkA{GRangesList}{GRangesList} classes\n",
       "in the \\pkg{GenomicRanges} package;\n",
       "\\item{} the \\LinkA{SummarizedExperiment}{SummarizedExperiment}\n",
       "class in the \\pkg{SummarizedExperiment} package;\n",
       "\\item{} the \\LinkA{GAlignments}{GAlignments},\n",
       "\\LinkA{GAlignmentPairs}{GAlignmentPairs},\n",
       "and \\LinkA{GAlignmentsList}{GAlignmentsList} classes\n",
       "in the \\pkg{GenomicAlignments} package;\n",
       "\\item{} the \\LinkA{TxDb}{TxDb} class in the\n",
       "\\pkg{GenomicFeatures} package;\n",
       "\\item{} the \\LinkA{BSgenome}{BSgenome} class in the\n",
       "\\pkg{BSgenome} package;\n",
       "\\item{} and more...\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\n",
       "\\end{enumerate}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Note}\\relax\n",
       "The full list of methods defined for a given generic function can be seen\n",
       "with e.g. \\code{showMethods(\"seqinfo\")} or \\code{showMethods(\"seqnames\")}\n",
       "(for the getters), and \\code{showMethods(\"seqinfo<-\")} or\n",
       "\\code{showMethods(\"seqnames<-\")} (for the setters a.k.a.\n",
       "\\emph{replacement methods}). Please be aware that this shows only methods\n",
       "defined in packages that are currently attached.\n",
       "\n",
       "The \\pkg{GenomicRanges} package defines \\code{seqinfo} and \\code{seqinfo<-}\n",
       "methods for these low-level data types: \\LinkA{List}{List} and\n",
       "\\LinkA{IntegerRangesList}{IntegerRangesList}. Those objects do not have the means\n",
       "to formally store sequence information. Thus, the wrappers simply store\n",
       "the \\LinkA{Seqinfo}{Seqinfo} object within \\code{metadata(x)}. Initially, the\n",
       "metadata is empty, so there is some effort to generate a reasonable\n",
       "default \\LinkA{Seqinfo}{Seqinfo}. The names of any \\LinkA{List}{List} are\n",
       "taken as the \\code{seqnames}, and the \\code{universe} of\n",
       "\\LinkA{IntegerRangesList}{IntegerRangesList} is taken as the \\code{genome}.\n",
       "\\end{Note}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "H. Pagès\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} The \\LinkA{seqlevelsStyle}{seqlevelsStyle} generic getter and setter for\n",
       "conveniently renaming the seqlevels of an object according\n",
       "to a given naming convention (e.g. NCBI or UCSC).\n",
       "\n",
       "\\item{} \\LinkA{Seqinfo}{Seqinfo} objects.\n",
       "\n",
       "\\item{} \\LinkA{GRanges}{GRanges} and \\LinkA{GRangesList}{GRangesList}\n",
       "objects in the \\pkg{GenomicRanges} package.\n",
       "\n",
       "\\item{} \\LinkA{SummarizedExperiment}{SummarizedExperiment} objects in the\n",
       "\\pkg{SummarizedExperiment} package.\n",
       "\n",
       "\\item{} \\LinkA{GAlignments}{GAlignments},\n",
       "\\LinkA{GAlignmentPairs}{GAlignmentPairs},\n",
       "and \\LinkA{GAlignmentsList}{GAlignmentsList} objects in the\n",
       "\\pkg{GenomicAlignments} package.\n",
       "\n",
       "\\item{} \\LinkA{TxDb}{TxDb} objects in the\n",
       "\\pkg{GenomicFeatures} package.\n",
       "\n",
       "\\item{} \\LinkA{BSgenome}{BSgenome} objects in the \\pkg{BSgenome} package.\n",
       "\n",
       "\\item{} \\LinkA{seqlevels-wrappers}{seqlevels.Rdash.wrappers} for convenience wrappers to the\n",
       "\\code{seqlevels} getter and setter.\n",
       "\n",
       "\\item{} \\code{\\LinkA{rankSeqlevels}{rankSeqlevels}}, on which \\code{sortSeqlevels} is\n",
       "based.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "## ---------------------------------------------------------------------\n",
       "## A. BASIC USAGE OF THE seqlevels() GETTER AND SETTER\n",
       "## ---------------------------------------------------------------------\n",
       "## Operations between 2 or more objects containing genomic ranges (e.g.\n",
       "## finding overlaps, comparing, or matching) only make sense if the\n",
       "## operands have the same seqlevels. So before performing such\n",
       "## operations, it is often necessary to adjust the seqlevels in\n",
       "## the operands so that they all have the same seqlevels. This is\n",
       "## typically done with the seqlevels() setter. The setter can be used\n",
       "## to rename, drop, add and/or reorder seqlevels of an object. The\n",
       "## examples below show how to mofify the seqlevels of a GRanges object\n",
       "## but the same would apply to any object containing sequence\n",
       "## information (i.e. with a seqinfo() component).\n",
       "library(GenomicRanges)\n",
       "\n",
       "gr <- GRanges(rep(c(\"chr2\", \"chr3\", \"chrM\"), 2), IRanges(1:6, 10))\n",
       "\n",
       "## Add new seqlevels:\n",
       "seqlevels(gr) <- c(\"chr1\", seqlevels(gr), \"chr4\")\n",
       "seqlevels(gr)\n",
       "seqlevelsInUse(gr)\n",
       "\n",
       "## Reorder existing seqlevels:\n",
       "seqlevels(gr) <- rev(seqlevels(gr))\n",
       "seqlevels(gr)\n",
       "\n",
       "## Drop all unused seqlevels:\n",
       "seqlevels(gr) <- seqlevelsInUse(gr)\n",
       "\n",
       "## Drop some seqlevels in use:\n",
       "seqlevels(gr, pruning.mode=\"coarse\") <- setdiff(seqlevels(gr), \"chr3\")\n",
       "gr\n",
       "\n",
       "## Rename, add, and reorder the seqlevels all at once:\n",
       "seqlevels(gr) <- c(\"chr1\", chr2=\"chr2\", chrM=\"Mitochondrion\")\n",
       "seqlevels(gr)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## B. DROP SEQLEVELS FROM A LIST-LIKE OBJECT\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "grl0 <- GRangesList(A=GRanges(\"chr2\", IRanges(3:2, 5)),\n",
       "                    B=GRanges(c(\"chr2\", \"chrMT\"), IRanges(7:6, 15)),\n",
       "                    C=GRanges(c(\"chrY\", \"chrMT\"), IRanges(17:16, 25)),\n",
       "                    D=GRanges())\n",
       "grl0\n",
       "\n",
       "grl1 <- grl0\n",
       "seqlevels(grl1, pruning.mode=\"coarse\") <- c(\"chr2\", \"chr5\")\n",
       "grl1  # grl0[[2]] was fully removed! (even if it had a range on chr2)\n",
       "\n",
       "## If what is desired is to remove the 2nd range in grl0[[2]] only (i.e.\n",
       "## the chrMT:6-15 range), or, more generally speaking, to remove the\n",
       "## ranges within each list element that are located on the seqlevels to\n",
       "## drop, then use pruning.mode=\"fine\" or pruning.mode=\"tidy\":\n",
       "grl2 <- grl0\n",
       "seqlevels(grl2, pruning.mode=\"fine\") <- c(\"chr2\", \"chr5\")\n",
       "grl2  # grl0[[2]] not removed, but chrMT:6-15 range removed from it\n",
       "\n",
       "## Like pruning.mode=\"fine\" but also removes grl0[[3]].\n",
       "grl3 <- grl0\n",
       "seqlevels(grl3, pruning.mode=\"tidy\") <- c(\"chr2\", \"chr5\")\n",
       "grl3\n",
       "\n",
       "library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)\n",
       "txdb <- TxDb.Dmelanogaster.UCSC.dm3.ensGene\n",
       "## Pruning mode \"coarse\" is particularly well suited on a GRangesList\n",
       "## object that contains exons grouped by transcript:\n",
       "ex_by_tx <- exonsBy(txdb, by=\"tx\")\n",
       "seqlevels(ex_by_tx)\n",
       "seqlevels(ex_by_tx, pruning.mode=\"coarse\") <- \"chr2L\"\n",
       "seqlevels(ex_by_tx)\n",
       "## Pruning mode \"tidy\" is particularly well suited on a GRangesList\n",
       "## object that contains transcripts grouped by gene:\n",
       "tx_by_gene <- transcriptsBy(txdb, by=\"gene\")\n",
       "seqlevels(tx_by_gene)\n",
       "seqlevels(tx_by_gene, pruning.mode=\"tidy\") <- \"chr2L\"\n",
       "seqlevels(tx_by_gene)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## C. RENAME THE SEQLEVELS OF A TxDb OBJECT\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)\n",
       "txdb <- TxDb.Dmelanogaster.UCSC.dm3.ensGene\n",
       "seqlevels(txdb)\n",
       "\n",
       "seqlevels(txdb) <- sub(\"chr\", \"\", seqlevels(txdb))\n",
       "seqlevels(txdb)\n",
       "\n",
       "seqlevels(txdb) <- paste0(\"CH\", seqlevels(txdb))\n",
       "seqlevels(txdb)\n",
       "\n",
       "seqlevels(txdb)[seqlevels(txdb) == \"CHM\"] <- \"M\"\n",
       "seqlevels(txdb)\n",
       "\n",
       "## Restore original seqlevels:\n",
       "seqlevels(txdb) <- seqlevels0(txdb)\n",
       "seqlevels(txdb)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## D. SORT SEQLEVELS IN \"NATURAL\" ORDER\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "sortSeqlevels(c(\"11\", \"Y\", \"1\", \"10\", \"9\", \"M\", \"2\"))\n",
       "\n",
       "seqlevels <- c(\"chrXI\", \"chrY\", \"chrI\", \"chrX\", \"chrIX\", \"chrM\", \"chrII\")\n",
       "sortSeqlevels(seqlevels)\n",
       "sortSeqlevels(seqlevels, X.is.sexchrom=TRUE)\n",
       "sortSeqlevels(seqlevels, X.is.sexchrom=FALSE)\n",
       "\n",
       "seqlevels <- c(\"chr2RHet\", \"chr4\", \"chrUextra\", \"chrYHet\",\n",
       "               \"chrM\", \"chrXHet\", \"chr2LHet\", \"chrU\",\n",
       "               \"chr3L\", \"chr3R\", \"chr2R\", \"chrX\")\n",
       "sortSeqlevels(seqlevels)\n",
       "\n",
       "gr <- GRanges()\n",
       "seqlevels(gr) <- seqlevels\n",
       "sortSeqlevels(gr)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## E. SUBSET OBJECTS BY SEQLEVELS\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "tx <- transcripts(txdb)\n",
       "seqlevels(tx)\n",
       "\n",
       "## Drop 'M', keep all others.\n",
       "seqlevels(tx, pruning.mode=\"coarse\") <- seqlevels(tx)[seqlevels(tx) != \"M\"]\n",
       "seqlevels(tx)\n",
       "\n",
       "## Drop all except 'ch3L' and 'ch3R'.\n",
       "seqlevels(tx, pruning.mode=\"coarse\") <- c(\"ch3L\", \"ch3R\")\n",
       "seqlevels(tx)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## F. FINDING METHODS\n",
       "## ---------------------------------------------------------------------\n",
       "\n",
       "showMethods(\"seqinfo\")\n",
       "showMethods(\"seqinfo<-\")\n",
       "\n",
       "showMethods(\"seqnames\")\n",
       "showMethods(\"seqnames<-\")\n",
       "\n",
       "showMethods(\"seqlevels\")\n",
       "showMethods(\"seqlevels<-\")\n",
       "\n",
       "if (interactive()) {\n",
       "  library(GenomicRanges)\n",
       "  ?`GRanges-class`\n",
       "}\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "seqinfo              package:GenomeInfoDb              R Documentation\n",
       "\n",
       "_\bA_\bc_\bc_\be_\bs_\bs_\bi_\bn_\bg/_\bm_\bo_\bd_\bi_\bf_\by_\bi_\bn_\bg _\bs_\be_\bq_\bu_\be_\bn_\bc_\be _\bi_\bn_\bf_\bo_\br_\bm_\ba_\bt_\bi_\bo_\bn\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     A set of generic functions for getting/setting/modifying the\n",
       "     sequence information stored in an object.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     seqinfo(x)\n",
       "     seqinfo(x,\n",
       "             new2old=NULL,\n",
       "             pruning.mode=c(\"error\", \"coarse\", \"fine\", \"tidy\")) <- value\n",
       "     \n",
       "     seqnames(x)\n",
       "     seqnames(x) <- value\n",
       "     \n",
       "     seqlevels(x)\n",
       "     seqlevels(x,\n",
       "               pruning.mode=c(\"error\", \"coarse\", \"fine\", \"tidy\")) <- value\n",
       "     sortSeqlevels(x, X.is.sexchrom=NA)\n",
       "     seqlevelsInUse(x)\n",
       "     seqlevels0(x)\n",
       "     \n",
       "     seqlengths(x)\n",
       "     seqlengths(x) <- value\n",
       "     \n",
       "     isCircular(x)\n",
       "     isCircular(x) <- value\n",
       "     \n",
       "     genome(x)\n",
       "     genome(x) <- value\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "       x: Any object containing sequence information i.e. with a\n",
       "          ‘seqinfo()’ component.\n",
       "\n",
       " new2old: The ‘new2old’ argument allows the user to rename, drop, add\n",
       "          and/or reorder the \"sequence levels\" in ‘x’.\n",
       "\n",
       "          ‘new2old’ can be ‘NULL’ or an integer vector with one element\n",
       "          per entry in Seqinfo object ‘value’ (i.e. ‘new2old’ and\n",
       "          ‘value’ must have the same length) describing how the \"new\"\n",
       "          sequence levels should be mapped to the \"old\" sequence\n",
       "          levels, that is, how the entries in ‘value’ should be mapped\n",
       "          to the entries in ‘seqinfo(x)’.  The values in ‘new2old’ must\n",
       "          be >= 1 and <= ‘length(seqinfo(x))’.  ‘NA’s are allowed and\n",
       "          indicate sequence levels that are being added.  Old sequence\n",
       "          levels that are not represented in ‘new2old’ will be dropped,\n",
       "          but this will fail if those levels are in use (e.g. if ‘x’ is\n",
       "          a GRanges object with ranges defined on those sequence\n",
       "          levels) unless a pruning mode is specified via the\n",
       "          ‘pruning.mode’ argument (see below).\n",
       "\n",
       "          If ‘new2old=NULL’, then sequence levels can only be added to\n",
       "          the existing ones, that is, ‘value’ must have at least as\n",
       "          many entries as ‘seqinfo(x)’ (i.e. ‘length(values) >=\n",
       "          length(seqinfo(x))’) and also\n",
       "          ‘seqlevels(values)[seq_len(length(seqlevels(x)))]’ must be\n",
       "          identical to ‘seqlevels(x)’.\n",
       "\n",
       "          Note that most of the times it's easier to proceed in 2\n",
       "          steps:\n",
       "\n",
       "           1. First align the seqlevels on the left (‘seqlevels(x)’)\n",
       "              with the seqlevels on the right.\n",
       "\n",
       "           2. Then call ‘seqinfo(x) <- value’. Because ‘seqlevels(x)’\n",
       "              and ‘seqlevels(value)’ now are identical, there's no need\n",
       "              to specify ‘new2old’.\n",
       "\n",
       "          This 2-step approach will typically look like this:\n",
       "\n",
       "              seqlevels(x) <- seqlevels(value)  # align seqlevels\n",
       "              seqinfo(x) <- seqinfo(value)  # guaranteed to work\n",
       "          \n",
       "          Or, if ‘x’ has seqlevels not in ‘value’, it will look like\n",
       "          this:\n",
       "\n",
       "              seqlevels(x, pruning.mode=\"coarse\") <- seqlevels(value)\n",
       "              seqinfo(x) <- seqinfo(value)  # guaranteed to work\n",
       "          \n",
       "          The ‘pruning.mode’ argument will control what happens to ‘x’\n",
       "          when some of its seqlevels get droppped. See below for more\n",
       "          information.\n",
       "\n",
       "pruning.mode: When some of the seqlevels to drop from ‘x’ are in use\n",
       "          (i.e. have ranges on them), the ranges on these sequences\n",
       "          need to be removed before the seqlevels can be dropped. We\n",
       "          call this _pruning_. The ‘pruning.mode’ argument controls how\n",
       "          to _prune_ ‘x’.  Four pruning modes are currently defined:\n",
       "          ‘\"error\"’, ‘\"coarse\"’, ‘\"fine\"’, and ‘\"tidy\"’. ‘\"error\"’ is\n",
       "          the default. In this mode, no pruning is done and an error is\n",
       "          raised. The other pruning modes do the following:\n",
       "\n",
       "            • ‘\"coarse\"’: Remove the elements in ‘x’ where the\n",
       "              seqlevels to drop are in use. Typically reduces the\n",
       "              length of ‘x’. Note that if ‘x’ is a list-like object\n",
       "              (e.g.  GRangesList, GAlignmentPairs, or GAlignmentsList),\n",
       "              then any list element in ‘x’ where at least one of the\n",
       "              sequence levels to drop is in use is _fully_ removed. In\n",
       "              other words, when ‘pruning.mode=\"coarse\"’, the\n",
       "              ‘seqlevels’ setter will keep or remove _full list\n",
       "              elements_ and not try to change their content. This\n",
       "              guarantees that the exact ranges (and their order) inside\n",
       "              the individual list elements are preserved. This can be a\n",
       "              desirable property when the list elements represent\n",
       "              compound features like exons grouped by transcript\n",
       "              (stored in a GRangesList object as returned by ‘exonsBy(\n",
       "              , by=\"tx\")’), or paired-end or fusion reads, etc...\n",
       "\n",
       "            • ‘\"fine\"’: Supported on list-like objects only. Removes\n",
       "              the ranges that are on the sequences to drop. This\n",
       "              removal is done within each list element of the original\n",
       "              object ‘x’ and doesn't affect its length or the order of\n",
       "              its list elements.  In other words, the pruned object is\n",
       "              guaranteed to be _parallel_ to the original object.\n",
       "\n",
       "            • ‘\"tidy\"’: Like the ‘\"fine\"’ pruning above but also\n",
       "              removes the list elements that become empty as the result\n",
       "              of the pruning. Note that this pruning mode is\n",
       "              particularly well suited on a GRangesList object that\n",
       "              contains transcripts grouped by gene, as returned by\n",
       "              ‘transcriptsBy( , by=\"gene\")’.  Finally note that, as a\n",
       "              convenience, this pruning mode is supported on non\n",
       "              list-like objects (e.g.  GRanges or GAlignments objects)\n",
       "              and, in this case, is equivalent to the ‘\"coarse\"’ mode.\n",
       "\n",
       "          See the \"B. DROP SEQLEVELS FROM A LIST-LIKE OBJECT\" section\n",
       "          in the examples below for an extensive illustration of these\n",
       "          pruning modes.\n",
       "\n",
       "   value: Typically a Seqinfo object for the ‘seqinfo’ setter.\n",
       "\n",
       "          Either a named or unnamed character vector for the\n",
       "          ‘seqlevels’ setter.\n",
       "\n",
       "          A vector containing the sequence information to store for the\n",
       "          other setters.\n",
       "\n",
       "X.is.sexchrom: A logical indicating whether X refers to the sexual\n",
       "          chromosome or to chromosome with Roman Numeral X. If ‘NA’,\n",
       "          ‘sortSeqlevels’ does its best to \"guess\".\n",
       "\n",
       "_\bI_\bt _\ba_\bl_\bl _\br_\be_\bv_\bo_\bl_\bv_\be_\bs _\ba_\br_\bo_\bu_\bn_\bd _\bS_\be_\bq_\bi_\bn_\bf_\bo _\bo_\bb_\bj_\be_\bc_\bt_\bs:\n",
       "\n",
       "     The Seqinfo class plays a central role for the functions described\n",
       "     in this man page because:\n",
       "\n",
       "       1. All these functions (except ‘seqinfo’, ‘seqlevelsInUse’, and\n",
       "          ‘seqlevels0’) work on a Seqinfo object.\n",
       "\n",
       "       2. For classes that implement it, the ‘seqinfo’ getter should\n",
       "          return a Seqinfo object.\n",
       "\n",
       "       3. Default ‘seqlevels’, ‘seqlengths’, ‘isCircular’, and ‘genome’\n",
       "          getters and setters are provided.  By default, ‘seqlevels(x)’\n",
       "          does ‘seqlevels(seqinfo(x))’, ‘seqlengths(x)’ does\n",
       "          ‘seqlengths(seqinfo(x))’, ‘isCircular(x)’ does\n",
       "          ‘isCircular(seqinfo(x))’, and ‘genome(x)’ does\n",
       "          ‘genome(seqinfo(x))’.  So any class with a ‘seqinfo’ getter\n",
       "          will have all the above getters work out-of-the-box. If, in\n",
       "          addition, the class defines a ‘seqinfo’ setter, then all the\n",
       "          corresponding setters will also work out-of-the-box.\n",
       "\n",
       "          Examples of containers that have a ‘seqinfo’ getter and\n",
       "          setter:\n",
       "\n",
       "            • the GRanges and GRangesList classes in the\n",
       "              ‘GenomicRanges’ package;\n",
       "\n",
       "            • the SummarizedExperiment class in the\n",
       "              ‘SummarizedExperiment’ package;\n",
       "\n",
       "            • the GAlignments, GAlignmentPairs, and GAlignmentsList\n",
       "              classes in the ‘GenomicAlignments’ package;\n",
       "\n",
       "            • the TxDb class in the ‘GenomicFeatures’ package;\n",
       "\n",
       "            • the BSgenome class in the ‘BSgenome’ package;\n",
       "\n",
       "            • and more...\n",
       "\n",
       "_\bN_\bo_\bt_\be:\n",
       "\n",
       "     The full list of methods defined for a given generic function can\n",
       "     be seen with e.g. ‘showMethods(\"seqinfo\")’ or\n",
       "     ‘showMethods(\"seqnames\")’ (for the getters), and\n",
       "     ‘showMethods(\"seqinfo<-\")’ or ‘showMethods(\"seqnames<-\")’ (for the\n",
       "     setters a.k.a.  _replacement methods_). Please be aware that this\n",
       "     shows only methods defined in packages that are currently\n",
       "     attached.\n",
       "\n",
       "     The ‘GenomicRanges’ package defines ‘seqinfo’ and ‘seqinfo<-’\n",
       "     methods for these low-level data types: List and\n",
       "     IntegerRangesList. Those objects do not have the means to formally\n",
       "     store sequence information. Thus, the wrappers simply store the\n",
       "     Seqinfo object within ‘metadata(x)’. Initially, the metadata is\n",
       "     empty, so there is some effort to generate a reasonable default\n",
       "     Seqinfo. The names of any List are taken as the ‘seqnames’, and\n",
       "     the ‘universe’ of IntegerRangesList is taken as the ‘genome’.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     H. Pagès\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "        • The seqlevelsStyle generic getter and setter for conveniently\n",
       "          renaming the seqlevels of an object according to a given\n",
       "          naming convention (e.g. NCBI or UCSC).\n",
       "\n",
       "        • Seqinfo objects.\n",
       "\n",
       "        • GRanges and GRangesList objects in the ‘GenomicRanges’\n",
       "          package.\n",
       "\n",
       "        • SummarizedExperiment objects in the ‘SummarizedExperiment’\n",
       "          package.\n",
       "\n",
       "        • GAlignments, GAlignmentPairs, and GAlignmentsList objects in\n",
       "          the ‘GenomicAlignments’ package.\n",
       "\n",
       "        • TxDb objects in the ‘GenomicFeatures’ package.\n",
       "\n",
       "        • BSgenome objects in the ‘BSgenome’ package.\n",
       "\n",
       "        • seqlevels-wrappers for convenience wrappers to the\n",
       "          ‘seqlevels’ getter and setter.\n",
       "\n",
       "        • ‘rankSeqlevels’, on which ‘sortSeqlevels’ is based.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## A. BASIC USAGE OF THE seqlevels() GETTER AND SETTER\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## Operations between 2 or more objects containing genomic ranges (e.g.\n",
       "     ## finding overlaps, comparing, or matching) only make sense if the\n",
       "     ## operands have the same seqlevels. So before performing such\n",
       "     ## operations, it is often necessary to adjust the seqlevels in\n",
       "     ## the operands so that they all have the same seqlevels. This is\n",
       "     ## typically done with the seqlevels() setter. The setter can be used\n",
       "     ## to rename, drop, add and/or reorder seqlevels of an object. The\n",
       "     ## examples below show how to mofify the seqlevels of a GRanges object\n",
       "     ## but the same would apply to any object containing sequence\n",
       "     ## information (i.e. with a seqinfo() component).\n",
       "     library(GenomicRanges)\n",
       "     \n",
       "     gr <- GRanges(rep(c(\"chr2\", \"chr3\", \"chrM\"), 2), IRanges(1:6, 10))\n",
       "     \n",
       "     ## Add new seqlevels:\n",
       "     seqlevels(gr) <- c(\"chr1\", seqlevels(gr), \"chr4\")\n",
       "     seqlevels(gr)\n",
       "     seqlevelsInUse(gr)\n",
       "     \n",
       "     ## Reorder existing seqlevels:\n",
       "     seqlevels(gr) <- rev(seqlevels(gr))\n",
       "     seqlevels(gr)\n",
       "     \n",
       "     ## Drop all unused seqlevels:\n",
       "     seqlevels(gr) <- seqlevelsInUse(gr)\n",
       "     \n",
       "     ## Drop some seqlevels in use:\n",
       "     seqlevels(gr, pruning.mode=\"coarse\") <- setdiff(seqlevels(gr), \"chr3\")\n",
       "     gr\n",
       "     \n",
       "     ## Rename, add, and reorder the seqlevels all at once:\n",
       "     seqlevels(gr) <- c(\"chr1\", chr2=\"chr2\", chrM=\"Mitochondrion\")\n",
       "     seqlevels(gr)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## B. DROP SEQLEVELS FROM A LIST-LIKE OBJECT\n",
       "     ## ---------------------------------------------------------------------\n",
       "     \n",
       "     grl0 <- GRangesList(A=GRanges(\"chr2\", IRanges(3:2, 5)),\n",
       "                         B=GRanges(c(\"chr2\", \"chrMT\"), IRanges(7:6, 15)),\n",
       "                         C=GRanges(c(\"chrY\", \"chrMT\"), IRanges(17:16, 25)),\n",
       "                         D=GRanges())\n",
       "     grl0\n",
       "     \n",
       "     grl1 <- grl0\n",
       "     seqlevels(grl1, pruning.mode=\"coarse\") <- c(\"chr2\", \"chr5\")\n",
       "     grl1  # grl0[[2]] was fully removed! (even if it had a range on chr2)\n",
       "     \n",
       "     ## If what is desired is to remove the 2nd range in grl0[[2]] only (i.e.\n",
       "     ## the chrMT:6-15 range), or, more generally speaking, to remove the\n",
       "     ## ranges within each list element that are located on the seqlevels to\n",
       "     ## drop, then use pruning.mode=\"fine\" or pruning.mode=\"tidy\":\n",
       "     grl2 <- grl0\n",
       "     seqlevels(grl2, pruning.mode=\"fine\") <- c(\"chr2\", \"chr5\")\n",
       "     grl2  # grl0[[2]] not removed, but chrMT:6-15 range removed from it\n",
       "     \n",
       "     ## Like pruning.mode=\"fine\" but also removes grl0[[3]].\n",
       "     grl3 <- grl0\n",
       "     seqlevels(grl3, pruning.mode=\"tidy\") <- c(\"chr2\", \"chr5\")\n",
       "     grl3\n",
       "     \n",
       "     library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)\n",
       "     txdb <- TxDb.Dmelanogaster.UCSC.dm3.ensGene\n",
       "     ## Pruning mode \"coarse\" is particularly well suited on a GRangesList\n",
       "     ## object that contains exons grouped by transcript:\n",
       "     ex_by_tx <- exonsBy(txdb, by=\"tx\")\n",
       "     seqlevels(ex_by_tx)\n",
       "     seqlevels(ex_by_tx, pruning.mode=\"coarse\") <- \"chr2L\"\n",
       "     seqlevels(ex_by_tx)\n",
       "     ## Pruning mode \"tidy\" is particularly well suited on a GRangesList\n",
       "     ## object that contains transcripts grouped by gene:\n",
       "     tx_by_gene <- transcriptsBy(txdb, by=\"gene\")\n",
       "     seqlevels(tx_by_gene)\n",
       "     seqlevels(tx_by_gene, pruning.mode=\"tidy\") <- \"chr2L\"\n",
       "     seqlevels(tx_by_gene)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## C. RENAME THE SEQLEVELS OF A TxDb OBJECT\n",
       "     ## ---------------------------------------------------------------------\n",
       "     \n",
       "     library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)\n",
       "     txdb <- TxDb.Dmelanogaster.UCSC.dm3.ensGene\n",
       "     seqlevels(txdb)\n",
       "     \n",
       "     seqlevels(txdb) <- sub(\"chr\", \"\", seqlevels(txdb))\n",
       "     seqlevels(txdb)\n",
       "     \n",
       "     seqlevels(txdb) <- paste0(\"CH\", seqlevels(txdb))\n",
       "     seqlevels(txdb)\n",
       "     \n",
       "     seqlevels(txdb)[seqlevels(txdb) == \"CHM\"] <- \"M\"\n",
       "     seqlevels(txdb)\n",
       "     \n",
       "     ## Restore original seqlevels:\n",
       "     seqlevels(txdb) <- seqlevels0(txdb)\n",
       "     seqlevels(txdb)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## D. SORT SEQLEVELS IN \"NATURAL\" ORDER\n",
       "     ## ---------------------------------------------------------------------\n",
       "     \n",
       "     sortSeqlevels(c(\"11\", \"Y\", \"1\", \"10\", \"9\", \"M\", \"2\"))\n",
       "     \n",
       "     seqlevels <- c(\"chrXI\", \"chrY\", \"chrI\", \"chrX\", \"chrIX\", \"chrM\", \"chrII\")\n",
       "     sortSeqlevels(seqlevels)\n",
       "     sortSeqlevels(seqlevels, X.is.sexchrom=TRUE)\n",
       "     sortSeqlevels(seqlevels, X.is.sexchrom=FALSE)\n",
       "     \n",
       "     seqlevels <- c(\"chr2RHet\", \"chr4\", \"chrUextra\", \"chrYHet\",\n",
       "                    \"chrM\", \"chrXHet\", \"chr2LHet\", \"chrU\",\n",
       "                    \"chr3L\", \"chr3R\", \"chr2R\", \"chrX\")\n",
       "     sortSeqlevels(seqlevels)\n",
       "     \n",
       "     gr <- GRanges()\n",
       "     seqlevels(gr) <- seqlevels\n",
       "     sortSeqlevels(gr)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## E. SUBSET OBJECTS BY SEQLEVELS\n",
       "     ## ---------------------------------------------------------------------\n",
       "     \n",
       "     tx <- transcripts(txdb)\n",
       "     seqlevels(tx)\n",
       "     \n",
       "     ## Drop 'M', keep all others.\n",
       "     seqlevels(tx, pruning.mode=\"coarse\") <- seqlevels(tx)[seqlevels(tx) != \"M\"]\n",
       "     seqlevels(tx)\n",
       "     \n",
       "     ## Drop all except 'ch3L' and 'ch3R'.\n",
       "     seqlevels(tx, pruning.mode=\"coarse\") <- c(\"ch3L\", \"ch3R\")\n",
       "     seqlevels(tx)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## F. FINDING METHODS\n",
       "     ## ---------------------------------------------------------------------\n",
       "     \n",
       "     showMethods(\"seqinfo\")\n",
       "     showMethods(\"seqinfo<-\")\n",
       "     \n",
       "     showMethods(\"seqnames\")\n",
       "     showMethods(\"seqnames<-\")\n",
       "     \n",
       "     showMethods(\"seqlevels\")\n",
       "     showMethods(\"seqlevels<-\")\n",
       "     \n",
       "     if (interactive()) {\n",
       "       library(GenomicRanges)\n",
       "       ?`GRanges-class`\n",
       "     }\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?seqlevels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a3f63963-db3a-4b3a-bdba-8bff62d5ad1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicScores\n",
      "\n",
      "\n",
      "Attaching package: ‘GenomicScores’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ShortRead’:\n",
      "\n",
      "    name\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    citation\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(phastCons100way.UCSC.hg38)\n",
    "phast <- phastCons100way.UCSC.hg38\n",
    "seqmask <- seqlevels(exonsbygene) %in% seqlevels(phast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fb03f4dc-7cb7-4193-9cec-4ab37b230e8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "29424"
      ],
      "text/latex": [
       "29424"
      ],
      "text/markdown": [
       "29424"
      ],
      "text/plain": [
       "[1] 29424"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "exonsbygene <- keepStandardChromosomes(exonsbygene, pruning.mode=\"coarse\")\n",
    "length(exonsbygene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fe5f08e0-f3bc-4882-8b24-d9162df15ce0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "557060"
      ],
      "text/latex": [
       "557060"
      ],
      "text/markdown": [
       "557060"
      ],
      "text/plain": [
       "[1] 557060"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allexons <- unlist(exonsbygene)\n",
    "length(allexons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "10a384d9-a687-4893-b4b0-f5910fa6c603",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 2 metadata columns:\n",
       "    seqnames            ranges strand |   exon_id   exon_name\n",
       "       <Rle>         <IRanges>  <Rle> | <integer> <character>\n",
       "  1    chr19 58345178-58347029      - |    585749        <NA>\n",
       "  1    chr19 58345183-58347029      - |    585750        <NA>\n",
       "  1    chr19 58346854-58347029      - |    585751        <NA>\n",
       "  1    chr19 58346858-58347029      - |    585752        <NA>\n",
       "  1    chr19 58346860-58347029      - |    585753        <NA>\n",
       "  1    chr19 58347353-58347634      - |    585754        <NA>\n",
       "  -------\n",
       "  seqinfo: 25 sequences (1 circular) from hg38 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(allexons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7be62a0a-8c88-4a1b-b456-40bc8f981017",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "assuming hg38 and Genome Reference Consortium GRCh38 represent the same genome build between query ranges and the GScores object, respectively.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 3 metadata columns:\n",
       "    seqnames            ranges strand |   exon_id   exon_name     phast\n",
       "       <Rle>         <IRanges>  <Rle> | <integer> <character> <numeric>\n",
       "  1    chr19 58345178-58347029      - |    585749        <NA> 0.0209503\n",
       "  1    chr19 58345183-58347029      - |    585750        <NA> 0.0204656\n",
       "  1    chr19 58346854-58347029      - |    585751        <NA> 0.0318182\n",
       "  1    chr19 58346858-58347029      - |    585752        <NA> 0.0325581\n",
       "  1    chr19 58346860-58347029      - |    585753        <NA> 0.0329412\n",
       "  1    chr19 58347353-58347634      - |    585754        <NA> 0.1191489\n",
       "  -------\n",
       "  seqinfo: 25 sequences (1 circular) from hg38 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allexons$phast <- gscores(phast, allexons)$default\n",
    "head(allexons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f1e2cb3a-a771-4953-bb39-4c694a9dc479",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'all_exons' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'all_exons' not found\nTraceback:\n",
      "1. head(all_exons)",
      "2. .handleSimpleError(function (cond) \n . .Internal(C_tryCatchHelper(addr, 1L, cond)), \"object 'all_exons' not found\", \n .     base::quote(head(all_exons)))",
      "3. h(simpleError(msg, call))"
     ]
    }
   ],
   "source": [
    "head(all_exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49654db8-fa4a-4a66-84ee-53389523554b",
   "metadata": {},
   "outputs": [],
   "source": [
    "check = data.frame("
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0021bce1-8727-4e32-8b52-d57265ebac0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "exonsbygene <- relist(allexons, exonsbygene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e5c3eef6-676e-4427-964e-4f0b8c1be691",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(org.Hs.eg.db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "83a0dadf-1c4d-44fc-bdb5-e0de74641038",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n"
     ]
    }
   ],
   "source": [
    "exonsbygene2 <- exonsbygene\n",
    "names(exonsbygene2) <- select(org.Hs.eg.db, keys=names(exonsbygene), columns=\"SYMBOL\")$SY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee89a033-9fa1-4710-af9e-db3077cc76ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "dtf <- data.frame(GeneSymbol=names(exonsbygene2),\n",
    "                  Conservation=tapply(allexons$phast, names(allexons), mean),\n",
    "                  stringsAsFactors=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "582d3a43-bb3e-4319-8a63-ecc106542559",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(dtf, \"/projects/ps-renlab/earmand/projects/species_comparison/analyses/gene_feature_distribution/average_exon_phastcons.tsv\", sep='\\t', row.names =FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bbf2aea2-1204-491f-bccf-28fbc0bbc751",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GeneSymbol</th><th scope=col>Conservation</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A1BG     </td><td>0.06276106</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>NAT2     </td><td>0.40282655</td></tr>\n",
       "\t<tr><th scope=row>100</th><td>ADA      </td><td>0.49428648</td></tr>\n",
       "\t<tr><th scope=row>1000</th><td>CDH2     </td><td>0.63736205</td></tr>\n",
       "\t<tr><th scope=row>100008586</th><td>GAGE12F  </td><td>0.07520020</td></tr>\n",
       "\t<tr><th scope=row>100009613</th><td>LINC02584</td><td>0.91239343</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & GeneSymbol & Conservation\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A1BG      & 0.06276106\\\\\n",
       "\t10 & NAT2      & 0.40282655\\\\\n",
       "\t100 & ADA       & 0.49428648\\\\\n",
       "\t1000 & CDH2      & 0.63736205\\\\\n",
       "\t100008586 & GAGE12F   & 0.07520020\\\\\n",
       "\t100009613 & LINC02584 & 0.91239343\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | GeneSymbol &lt;chr&gt; | Conservation &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1 | A1BG      | 0.06276106 |\n",
       "| 10 | NAT2      | 0.40282655 |\n",
       "| 100 | ADA       | 0.49428648 |\n",
       "| 1000 | CDH2      | 0.63736205 |\n",
       "| 100008586 | GAGE12F   | 0.07520020 |\n",
       "| 100009613 | LINC02584 | 0.91239343 |\n",
       "\n"
      ],
      "text/plain": [
       "          GeneSymbol Conservation\n",
       "1         A1BG       0.06276106  \n",
       "10        NAT2       0.40282655  \n",
       "100       ADA        0.49428648  \n",
       "1000      CDH2       0.63736205  \n",
       "100008586 GAGE12F    0.07520020  \n",
       "100009613 LINC02584  0.91239343  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(dtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30c6c8cf-1451-4dee-97dc-6eb436d465bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:circle_plot]",
   "language": "R",
   "name": "conda-env-circle_plot-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
